### ISSUE TITLE
Adaptive resources when injecting sidecars

### ISSUE THREAD

**Issue by erenliu168** [istio/istio#48623]
**Describe the feature request**

I want to set resources (cpu and memory) adaptively, such as setting sidecar resources (cpu and memory) according to the resources of pod and so on.

**Describe alternatives you've considered**

1. Set some factors or flags in MeshConfig for resource levels.
2. Enhance the injector template, for example adding some annotations. It will bring much work when there are thousands of pods in k8s so it must be used together with global settings.

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[ ] Networking
[X] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by zirain** [comment]
are you looking for `sidecar.istio.io/proxyCPU`?

**Comment by erenliu168** [comment]
> are you looking for `sidecar.istio.io/proxyCPU`?

Yes. But is dosen't satisfy my requirments. I don't know the pod resources of users in advance.

**Comment by hzxuzhonghu** [comment]
How do you want to map the resources between sidecar and app container?

**Comment by erenliu168** [comment]
> How do you want to map the resources between sidecar and app container?

Hi, there. I suppose there are three ways to implement.

1. just one factor in MeshConfig
resource of sidecar: 
**(resource of app container)\*factor**        (both request and limit of both cpu and mem )
This way is very simple and annotation in injection template (sidecar.istio.io/proxyCPU  etc.) can override it.

2. two factors  request_factor and limit_factor in MeshConfig
request of sidecar: 
**(request of app container)\*request_factor**        (both cpu and mem)
limit of sidecar: 
**(limit of app container)\*limit_factor**        (both cpu and mem)
and annotation in injection template (sidecar.istio.io/proxyCPU  etc.) can override it.

3. resource_level->request_factor,limit_factor map and default_level in MeshConfig, and level in injection template annotation to override
request of sidecar: 
**(request of app container)\*map[default_level ][0]**        (both cpu and mem)
limit of sidecar: 
**(limit of app container)\*map[default_level ][1]**        (both cpu and mem)
or there is annotation (may be sidecar.istio.io/resourceLevel) in pod template,  it can be
request of sidecar: 
**(request of app container)\*map[pod resourceLevel][0]**        (both cpu and mem)
limit of sidecar: 
**(limit of app container)\*map[pod resourceLevel][1]**   (both cpu and mem)
It can alse be overridden by existed anntations (sidecar.istio.io/proxyCPU  etc.)

------------------------------------------------------
In above ways we can also **split the factor to cpu factor and mem factor**. 
In this situation, the resource will be (for example just suppose the map is resource_level
->cpu_request_factor,cpu_limit_factor,mem_request_factor,mem_limit_factor in the 3rd way.)

cpu request of sidecar: 
**(cpu request of app container)\*map[pod resourceLevel][0]**
cpu limit of sidecar: 
**(cpu limit of app container)\*map[pod resourceLevel][1]**
memory request of sidecar: 
**(memory request of app container)\*map[pod resourceLevel][2]**
memory limit of sidecar: 
**(memory limit of app container)\*map[pod resourceLevel][3]**

Any good advice? 3ks

**Comment by howardjohn** [comment]
I don't think we should do this. This is quite complex calculation given there can be many different app containers, some may be added after istio's dynamically, there is many cases, and its very often that sidecar resource **usage** has no correlation to app resource **requests**.

I would recommend instead solving this out of band by dynamically setting the resource annotations (In CI, a controller, whatever)
